pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,385 @@
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	obj, found, err = p.cache.GetByKey(globalCacheKey)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	return nil, false, nil
+}
+
+// Plugin is the interface calling ExecPlugin. This is mainly for testability
+// so tests don't have to actually exec any processes.
+type Plugin interface {
+	ExecPlugin(ctx context.Context, image string) (*credentials.CredentialProviderResponse, error)
+}
+
+// execPlugin is the implementation of the Plugin interface that execs a credential provider plugin based
+// on it's name provided in CredentialProviderConfig. It is assumed that the executable is available in the
+// plugin directory provided by the kubelet.
+type execPlugin struct {
+	name         string
+	apiVersion   string
+	args         []string
+	envVars      []kubeletconfig.ExecEnvVar
+	pluginBinDir string
+}
+
+// ExecPlugin executes the plugin binary with arguments and environment variables specified in CredentialProviderConfig:
+//
+//  $ ENV_NAME=ENV_VALUE <plugin-name> args[0] args[1] <<<request
+//
+// The plugin is expected to receive the CredentialProviderRequest API via stdin from the kubelet and
+// return CredentialProviderResponse via stdout.
+func (e *execPlugin) ExecPlugin(ctx context.Context, image string) (*credentials.CredentialProviderResponse, error) {
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	stdin := &bytes.Buffer{}
+
+	// Use a catch-all timeout of 1 minute for all exec-based plugins, this should leave enough
+	// head room in case a plugin needs to retry a failed request while ensuring an exec plugin
+	// does not run forever. In the future we may want this timeout to be tweakable from the plugin
+	// config file.
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	pluginBin := filepath.Join(e.pluginBinDir, e.name)
+	cmd := exec.CommandContext(ctx, pluginBin, e.args...)
+	cmd.Stdout, cmd.Stderr, cmd.Stdin = stdout, stderr, stdin
+
+	cmd.Env = []string{}
+	for _, envVar := range e.envVars {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", envVar.Name, envVar.Value))
+	}
+
+	authRequest := &credentials.CredentialProviderRequest{Image: image}
+	data, err := e.encodeRequest(authRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode auth request: %v", err)
+	}
+
+	if _, err := stdin.Write(data); err != nil {
+		return nil, fmt.Errorf("error writing to stdin: %v", err)
+	}
+
+	err = cmd.Run()
+	if ctx.Err() != nil {
+		return nil, fmt.Errorf("error execing credential provider plugin %s for image %s: %w", e.name, image, ctx.Err())
+	}
+
+	if err != nil {
+		klog.V(2).Infof("Error execing credential provider plugin, stderr: %v", stderr.String())
+		return nil, fmt.Errorf("error execing credential provider plugin %s for image %s: %w", e.name, image, err)
+	}
+
+	response, err := e.decodeResponse(stdout.Bytes())
```

Comment:
should we verify the plugin responded with the apiVersion we expect?

Reply:
> should we verify the plugin responded with the apiVersion we expect?

I was thinking it shouldn't matter since we will always decode it into the internal type and the decoder will error if it sees an apiVersion it doesn't recognize.

Reply:
I guess that's true...

Reply:
I don't actually know how that erorr would surface though so I can add an explicit check to decode the apiVersion first and erorr if it doesn't match.